File: Semester3/ProgrammingParadigms/Lab3/OCaml/main.ml
let print_list lst =
  Printf.printf "[";
  List.iter (fun x -> Printf.printf "%d; " x) lst;
  Printf.printf "]\n"

(*1*)

let rec my_map f lst =
  match lst with
  | [] -> []
  | head :: tail -> (f head) :: my_map f tail



  let increment x = x + 1;;
  let input_list1 = [1; 2; 3; 4; 5];;
  
  let result1 = my_map increment input_list1;;

  print_list result1;;
  (*2*)

  let rec my_filter lst pred =
    match lst with
    | [] -> []
    | head :: tail when pred head -> head :: my_filter tail pred
    | _ :: tail -> my_filter tail pred


  let pred x = x mod 2 = 0;;

  let input_list2 = [1;2;3;4;5;6;7;8;9];;

  let result2 = my_filter input_list2 pred;;

  print_list result2;;

  (*3*)

  let rec my_reduce lst op acc =
    match lst with
    | [] -> acc
    | head :: tail -> my_reduce tail op (op head acc)
  
  let op x y = x + y
  
  let input_list3 = [1; 2; 3; 4; 5];;
  let result3 = my_reduce input_list3 op 0;;

  print_int result3;;
  Printf.printf("\n")

  (*4*)

  let my_average lst =
    float_of_int (my_reduce lst (fun x y -> x + y) 0) /. float_of_int (List.length lst)
  
  let input_list4 = [1;2;3;4;5;6;7;8;9];;
  let result4 = my_average input_list4;;

  print_float result4;;
  Printf.printf("\n")

  (*5*)

  let rec generate_acronym_recursive str =
    let rec aux acc = function
      | [] -> String.concat "" (List.rev acc)
      | word :: rest ->
          match String.trim word with
          | "" -> aux acc rest
          | s -> aux ((String.make 1 (Char.uppercase_ascii (String.get s 0))) :: acc) rest
    in
    aux [] (String.split_on_char ' ' str)
  ;;
  
  let input_string = "zaklad ubezpieczen spolecznych";;
  let acronym = generate_acronym_recursive input_string;;
  print_endline acronym;; 

  (*6*)

  let rec reduce_list myList = 
    let cubeValue value = value*value*value in 
    let sum = my_reduce myList (fun x y -> x + y) 0 in 
    match myList with
    | [] -> []
    | head :: tail when cubeValue head < sum -> head :: reduce_list tail
    | head :: tail -> reduce_list tail

  let input_list6 = [1;2;3;4;5;6;7;8;9];;
  let result6 = reduce_list input_list6;;

  print_list result6

File: Semester3/ProgrammingParadigms/Lab3/Scala/main.scala
import scala.languageFeature.reflectiveCalls
object app extends App {


    // 0

    println("Task 0: ")

    def Log(prefix: String)(datetime: String)(text: String): String = {
        "[" + prefix + "] " + datetime + " " + text
    }

    val logMessage = Log("WARN")("2023-10-28")("Hello, world!")
    println(logMessage)


    // 1

    println("Task 1: ")

    def Map(myList: List[Int], unaryFunction: Int => Int): List[Int] = {
        myList match {
            case Nil => Nil 
            case head :: tail => unaryFunction(head) :: Map(tail, unaryFunction) 
        }
    }

    val myIntegers = List(1, 2, 3, 4, 5)

    val doubledList = Map(myIntegers, (x: Int) => x * x)

    println(doubledList)


    // 2

    println("Task 2: ")

    def Filter(myList: List[Int], pred: Int => Boolean): List[Int] = {
        myList match {
            case Nil => Nil
            case head :: tail if pred(head) => head :: Filter(tail, pred)
            case head :: tail => Filter(tail, pred)
        }
    }

    val myIntegers2 = List(1,2,3,4,5,6,7,8,9,10)

    val result2 = Filter(myIntegers2, (x: Int) => x % 2 == 0)

    println(result2)

    // 3

    println("Task 3: ")

    def Reduce(myList: List[Int], op: (Int, Int) => Int, acc: Int): Int = {
        myList match {
            case Nil => acc
            case head :: tail => Reduce(tail, op, op(head, acc))
        }
    }

    val myIntegers3 = List(1,2,3)

    val result3 = Reduce(myIntegers3, (x,y) => x+y, 0)

    print(result3)

    // 4

    println("Task 4: ")

    def Average(myList: List[Int]): Float = {
        Reduce(myList, (x,y) => x+y, 0) / myList.length
    }

    val myIntegers4 = List(1,2,3,4,5,6,7,8,9,10)

    val result4 = Average(myIntegers4)

    print("Average of this list is: " + result4)


    // 5

    println("Task 5: ")


    def generateAcronym(myString: String): String = {
        val words = myString.split(" ")  
        val acronym = words.map(_.head.toUpper).mkString  
        acronym
    }

    val inputString = "Zakład Ubezpieczeń Społecznych"
    val acronym = generateAcronym(inputString)

    println(acronym)  

    // 6

    def reduceList(myList: List[Int]): List[Int] = {
        val sum = myList.sum
        def cubeValue(value: Int): Int = value * value * value
        myList match {
            case Nil => Nil
            case head :: tail if cubeValue(head) <= sum => head :: reduceList(tail)
            case _ :: tail => reduceList(tail)
        }
    }

    val myIntegers6 = List(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
    val result6 = reduceList(myIntegers6)
    println(result6)






}